File: Explanation_Engine/explanation_module/explanation_module.py
from typing import List, Dict, Any
from datetime import datetime

from .config import Config
from .fact_extractor import FactExtractor
from .llm_rewriter import LLMRewriter
from .product_card_builder import ProductCardBuilder

class ShopGenieExplanationModule:
    def __init__(self, config=None):
        self.config = config or Config()
        self.fact_extractor = FactExtractor(self.config)
        self.llm_rewriter = LLMRewriter(self.config)
        self.card_builder = ProductCardBuilder(self.config)
        self.stats = {"total": 0, "llm_success": 0}
    
    def process(self, query: str, ranked_products: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Process multiple ranked products (5-6 items)
        
        Args:
            query: User's search query
            ranked_products: List of RankedProduct objects (typically 5-6 items)
        
        Returns:
            Dict with chat response, product cards, and suggestions
        """
        
        # Limit to top 6 products maximum
        products_to_process = ranked_products[:6]
        total_received = len(ranked_products)
        total_processing = len(products_to_process)
        
        print(f"📊 Processing {total_processing} products (received {total_received})")
        
        cards = []
        for rank, rp in enumerate(products_to_process, 1):
            print(f"   🔍 Processing product #{rank}: {rp.get('product', {}).get('title', 'Unknown')[:40]}...")
            
            facts = self.fact_extractor.extract_facts(rp)
            bullet_str = self.fact_extractor.to_bullet_string(facts)
            
            llm_text = None
            if self.config.USE_LLM:
                product = rp.get('product', {})
                llm_text = self.llm_rewriter.rewrite(
                    product.get('title', ''),
                    product.get('platform', ''),
                    bullet_str
                )
            
            card = self.card_builder.build(rp, rank, facts, llm_text)
            cards.append(card)
            
            self.stats["total"] += 1
            if llm_text:
                self.stats["llm_success"] += 1
        
        # Build chat response with count of products
        chat_response = self._build_chat_response(query, cards, total_processing)
        
        return {
            "query": query,
            "timestamp": datetime.now().isoformat(),
            "total_products": total_processing,
            "total_found": total_received,
            "chat_response": chat_response,
            "product_cards": cards,
            "suggested_followups": self._get_followups(cards, query),
            "stats": self._get_stats()
        }
    
    def _build_chat_response(self, query: str, cards: List[Dict], total: int) -> str:
        """Build chat response with multiple products highlighted"""
        
        if not cards:
            return f"Sorry, no products found for '{query}'. Try a different search!"
        
        # Opening based on number of products
        if total == 1:
            opening = f"🔍 I found 1 great option for '{query}'!"
        else:
            opening = f"🔍 I found {total} great options for '{query}'!"
        
        response = opening + "\n\n"
        
        # Top pick highlight
        top = cards[0]
        response += f"🏆 **Top pick**: {top['name']}\n"
        response += f"   • {top['price']['display']} | ⭐ {top['rating']['display']}\n"
        
        if top.get('matchingFeatures'):
            response += f"   • Best for: {', '.join(top['matchingFeatures'][:2])}\n"
        
        if top['explanation']['natural']:
            response += f"\n✨ {top['explanation']['natural']}\n"
        
        # Brief of other products if more than 1
        if len(cards) > 1:
            response += f"\n📋 Also check out:\n"
            for card in cards[1:3]:  # Show next 2 products
                response += f"   • {card['name'][:35]} — {card['price']['display']} | ⭐ {card['rating']['display']}\n"
        
        response += f"\n💡 Scroll down to see all {total} recommendations with detailed product cards!"
        
        return response
    
    def _get_followups(self, cards: List[Dict], query: str) -> List[str]:
        """Generate smarter follow-up questions based on products"""
        
        followups = []
        
        # General follow-ups
        if len(cards) >= 2:
            followups.append(f"Compare the top 2 products")
        
        followups.append(f"Show me cheaper options")
        
        # Feature-based follow-ups from top product
        if cards and cards[0].get('matchingFeatures'):
            main_feature = cards[0]['matchingFeatures'][0]
            followups.append(f"Tell me more about {main_feature}")
        
        # Price-based follow-up
        prices = [c['price']['amount'] for c in cards if c.get('price', {}).get('amount')]
        if prices and len(prices) > 1:
            followups.append(f"What's the best value under ${min(prices) + 50:.0f}?")
        
        # Platform-based follow-up
        platforms = set(c.get('platform', {}).get('name', '') for c in cards)
        if len(platforms) > 1:
            followups.append(f"Show only {list(platforms)[0]} products")
        
        return followups[:5]  # Max 5 suggestions
    
    def _get_stats(self) -> Dict:
        total = max(1, self.stats["total"])
        return {
            "products_processed": self.stats["total"],
            "llm_success_count": self.stats["llm_success"],
            "llm_success_rate": f"{(self.stats['llm_success'] / total * 100):.1f}%"
        }

File: Explanation_Engine/explanation-module/explanationModule.ts
// backend/explanation-module/explanationModule.ts
import { RankedProduct, ExplanationOutput } from './types';
import { FactExtractor } from './factExtractor';
import { LLMRewriter } from './llmRewriter';
import { ProductCardBuilder } from './productCardBuilder';
import { config } from './config';

export class ShopGenieExplanationModule {
  private factExtractor: FactExtractor;
  private llmRewriter: LLMRewriter;
  private cardBuilder: ProductCardBuilder;
  private stats = { total: 0, llmSuccess: 0 };

  constructor() {
    this.factExtractor = new FactExtractor();
    this.llmRewriter = new LLMRewriter();
    this.cardBuilder = new ProductCardBuilder();
  }

  async process(query: string, rankedProducts: RankedProduct[]): Promise<ExplanationOutput> {
    const productsToProcess = rankedProducts.slice(0, 6);
    const totalReceived = rankedProducts.length;

    console.log(`📊 Processing ${productsToProcess.length} products (received ${totalReceived})`);

    const productCards = [];
    for (let i = 0; i < productsToProcess.length; i++) {
      const rp = productsToProcess[i];
      const rank = i + 1;
      console.log(`   🔍 Processing product #${rank}: ${rp.product.title.substring(0, 40)}...`);

      const facts = this.factExtractor.extractFacts(rp);
      const bulletString = this.factExtractor.toBulletString(facts);

      let llmText: string | null = null;
      if (config.useLLM) {
        llmText = await this.llmRewriter.rewrite(rp.product.title, rp.product.platform, bulletString);
      }

      const card = this.cardBuilder.build(rp, rank, facts, llmText);
      productCards.push(card);

      this.stats.total++;
      if (llmText) this.stats.llmSuccess++;
    }

    const chatResponse = this.buildChatResponse(query, productCards);
    const suggestedFollowups = this.getFollowups(productCards, query);

    return {
      query,
      timestamp: new Date().toISOString(),
      totalProducts: productCards.length,
      totalFound: totalReceived,
      chatResponse,
      productCards,
      suggestedFollowups,
      stats: {
        productsProcessed: this.stats.total,
        llmSuccessCount: this.stats.llmSuccess,
        llmSuccessRate: `${((this.stats.llmSuccess / Math.max(1, this.stats.total)) * 100).toFixed(1)}%`,
      },
    };
  }

  private buildChatResponse(query: string, cards: ExplanationOutput['productCards']): string {
    if (cards.length === 0) {
      return `Sorry, no products found for '${query}'. Try a different search!`;
    }

    let response = cards.length === 1
      ? `🔍 I found 1 great option for '${query}'!\n\n`
      : `🔍 I found ${cards.length} great options for '${query}'!\n\n`;

    const top = cards[0];
    response += `🏆 **Top pick**: ${top.name}\n`;
    response += `   • ${top.price.display} | ⭐ ${top.rating.display}\n`;
    if (top.keyFeatures.length) {
    response += `   • Features: ${top.keyFeatures.slice(0, 2).join(', ')}\n`;
    }
    if (top.explanation.natural) {
      response += `\n✨ ${top.explanation.natural}\n`;
    }

    if (cards.length > 1) {
      response += `\n📋 Also check out:\n`;
      for (let i = 1; i < Math.min(3, cards.length); i++) {
        const card = cards[i];
        response += `   • ${card.name.substring(0, 35)} — ${card.price.display} | ⭐ ${card.rating.display}\n`;
      }
    }

    response += `\n💡 Scroll down to see all ${cards.length} recommendations with detailed product cards!`;
    return response;
  }

  private getFollowups(cards: ExplanationOutput['productCards'], query: string): string[] {
    const followups: string[] = [];
    if (cards.length >= 2) followups.push('Compare the top 2 products');
    followups.push('Show me cheaper options');
    if (cards[0]?.keyFeatures.length) {
      followups.push(`Tell me more about ${cards[0].keyFeatures[0]}`);
    }
    const prices = cards.map(c => c.price.amount).filter(p => p > 0);
    if (prices.length > 1) {
      const minPrice = Math.min(...prices);
      followups.push(`What's the best value under $${Math.floor(minPrice + 50)}?`);
    }
    const platforms = new Set(cards.map(c => c.platform.name));
    if (platforms.size > 1) {
      followups.push(`Show only ${Array.from(platforms)[0]} products`);
    }
    return followups.slice(0, 5);
  }
}